lib/patientFinder/findPatientsIDs.rb

```diff
@@ -0,0 +1,203 @@
+			if(masterRecord.gender==testRecord.gender && masterRecord.ethnicity[:code]==testRecord.ethnicity[:code] && masterRecord.race[:code] == testRecord.race[:code])
+				#go inside the sections and collect codes and find if the codes from the test are ALL contained in the the same section of the master 
+				allergiesMatch=Utils::matchCodes(testRecord.allergies,masterRecord.allergies)
+				care_goalsMatch=Utils::matchCodes(testRecord.care_goals,masterRecord.care_goals)
+				conditionsMatch=Utils::matchCodes(testRecord.conditions,masterRecord.conditions)
+				encountersMatch=Utils::matchCodes(testRecord.encounters,masterRecord.encounters)
+				immunizationsMatch=Utils::matchCodes(testRecord.immunizations,masterRecord.immunizations)
+				medical_equipmentMatch=Utils::matchCodes(testRecord.medical_equipment,masterRecord.medical_equipment)
+				medicationsMatch=Utils::matchCodes(testRecord.medications,masterRecord.medications)
+				proceduresMatch=Utils::matchCodes(testRecord.procedures,masterRecord.procedures)
+				insurance_providersMatch=Utils::matchCodes(testRecord.insurance_providers,masterRecord.insurance_providers)
+
+				#create candidates based on matching codes - if all the above match then consider this master as candidate
+				if(allergiesMatch && care_goalsMatch && conditionsMatch && encountersMatch &&  immunizationsMatch && medical_equipmentMatch &&  medicationsMatch && proceduresMatch &&  insurance_providersMatch)
+					candidates<< masterRecord
+				end
+			end	
+		end	
+		candidates
+	end
+
+	#a utility that loads into mongoDB records from a zip containing QRDA files 
+	def self.collectQRDAPatientsFromZip(zipFile)
+		patientsFromQRDAs = Array.new
+		zipFilex = open(zipFile)
+		Zip::ZipFile.open(zipFilex.path) do |zipfile|
+			zipfile.entries.each do |entry|
+				data =zipfile.read(entry.name)
+				
+				doc = Nokogiri::XML(data)
+				doc.root.add_namespace_definition('cda', 'urn:hl7-org:v3')
+				doc.root.add_namespace_definition('sdtc', 'urn:hl7-org:sdtc')
+				patient_data = HealthDataStandards::Import::Cat1::PatientImporter.instance.parse_cat1(doc)
+				patient = Record.update_or_create(patient_data)
+				patientFromQRDA = PatientFromQRDA.new(entry.name, patient)
+				patientsFromQRDAs << patientFromQRDA
+			end
+		end
+		patientsFromQRDAs
+	end
+
+	def self.createCodeHash(thisEntry, patientCodesHash)
+		thisEntry.codes.each_pair do |code_set, coded_values|
+			coded_values.each do |coded_value|
+				patientCodesHash[coded_value]=code_set
+			end
+		end
+	end
+
+	#identifies if all the codes in a particular section of the test are in the same section of the master
+	def self.matchCodes(testEntries,masterEntries)
+		result = false
+
+		testCodesHash= Hash.new 
+		testEntries.each do |thisEntry|
+			createCodeHash(thisEntry, testCodesHash)
+		end
+
+		masterCodesHash= Hash.new 
+		masterEntries.each do |thisEntry|
+			createCodeHash(thisEntry, masterCodesHash)
+		end
+
+		if (testCodesHash.to_a - masterCodesHash.to_a).empty?
```

Comment:
This is more of a preference than anything else, but usually people will just return the value of `(testCodesHash.to_a - masterCodesHash.to_a).empty? ` rather than creating a dummy variable